### ISSUE TITLE
[BUG]Receive GoneException instead RequestTimeoutException in a timeout situation

### ISSUE THREAD

**Issue by Horcrux7** [Azure/azure-sdk-for-java#19845]
If internally an request timeout occur because to many parallel requests was send then I receive a GoneException. I expected a RequestTimeoutException. I use com.azure:azure-cosmos:4.12.0.

The follow sample code demonstrate it:

```
import com.azure.cosmos.CosmosAsyncClient;
import com.azure.cosmos.CosmosAsyncContainer;
import com.azure.cosmos.CosmosAsyncDatabase;
import com.azure.cosmos.CosmosClientBuilder;
import com.azure.cosmos.models.CosmosContainerProperties;

import reactor.core.publisher.Mono;

public class TestCosmosBug {

    public static void main( String[] args ) {
        // connect to the emulator
        CosmosClientBuilder builder = new CosmosClientBuilder();
        builder.endpoint( "https://cosmos.company.local:8081/" );
        builder.key( "C2y6yDjf5/R+ob0N8A7Cgv30VRDJIWEHLM+4QDU5DE2nQ9nDuVTqobD4b8mGGyPMbIZnqyMsEcaGQy67XIw/Jw==" );
        CosmosAsyncClient client = builder.buildAsyncClient();
        try {
            try {
                client.getDatabase( "testdb" ).delete().block(); // cleanup previous test
            } catch( Exception e ) {
                // ignore;
            }

            //create or get container
            client.createDatabaseIfNotExists( "testdb" ).block();
            CosmosContainerProperties containerProps = new CosmosContainerProperties( "testcolection", "/id" );
            CosmosAsyncDatabase database = client.getDatabase( "testdb" );
            database.createContainerIfNotExists( containerProps ).block();
            CosmosAsyncContainer container = database.getContainer( "testcolection" );

            byte[] data = new byte[250_000];
            Mono<?> mono = Mono.empty();

            // increment the count until an error occur. The mono block must be wait for more as 5 seconds that this error occur. 
            for( int i = 0; i < 1000; i++ ) {
                POJO pojo = new POJO();
                pojo.id = "id" + i;
                pojo.data = data;
                mono = mono.and( container.createItem( pojo ) );
            }

            //wait on write finish, it should be a RequestTimeoutException but is a GoneException. Occur after 5 seconds.
            mono.block();
        } catch( Throwable ex ) {
            System.err.println( ex.getClass() );
            System.err.println();
            ex.printStackTrace();
        }
        //shutdown
        client.close();
        System.exit( 0 );
    }

    private static class POJO {
        public String id;
        public byte[] data;
    }
}

```

**Comment by kushagraThapar** [comment]
@Horcrux7 - can you please provide the full stack trace ?

**Comment by Horcrux7** [comment]
```
{"ClassName":"GoneException","userAgent":"azsdk-java-cosmos/4.12.0 Windows10/10.0 JRE/11.0.7","statusCode":410,"resourceAddress":"rntbd://10.10.10.16:10253/apps/DocDbApp/services/DocDbServer21/partitions/a4cb4961-38c8-11e6-8106-8cdcd42c33be/replicas/1p/","innerErrorMessage":"AsyncRntbdRequestRecord(
{\"args\":{\"transportRequestId\":5,\"activityId\":\"8a37e431-867b-11eb-92fa-8b46417fc6ba\",\"origin\":\"rntbd://10.10.10.16:10253\",\"replicaPath\":\"/apps/DocDbApp/services/DocDbServer21/partitions/a4cb4961-38c8-11e6-8106-8cdcd42c33be/replicas/1p\",\"timeCreated\":\"2021-03-16T17:17:52.531586400Z\
",\"lifetime\":\"PT5.1371635S\"},\"requestLength\":333714,\"responseLength\":-1,\"status\":{\"done\":false,\"cancelled\":false,\"completedExceptionally\":false},\"timeline\":[{\"eventName\":\"created\",\"startTimeUTC\":\"2021-03-16T17:17:52.531586400Z\",\"durationInMicroSec\":0},{\"eventName\":\"que
ued\",\"startTimeUTC\":\"2021-03-16T17:17:52.531586400Z\",\"durationInMicroSec\":0},{\"eventName\":\"channelAcquisitionStarted\",\"startTimeUTC\":\"2021-03-16T17:17:52.531586400Z\",\"durationInMicroSec\":68816},{\"eventName\":\"pipelined\",\"startTimeUTC\":\"2021-03-16T17:17:52.600402800Z\",\"durati
onInMicroSec\":565488},{\"eventName\":\"transitTime\",\"startTimeUTC\":\"2021-03-16T17:17:53.165891600Z\",\"durationInMicroSec\":4503111},{\"eventName\":\"received\",\"startTimeUTC\":null,\"durationInMicroSec\":0},{\"eventName\":\"completed\",\"startTimeUTC\":null,\"durationInMicroSec\":0}]})","caus
eInfo":null,"responseHeaders":"{}","requestHeaders":"[Accept=application/json, x-ms-date=Tue, 16 Mar 2021 17:17:52 GMT, x-ms-documentdb-collection-rid=KbF7AOzF+og=, x-ms-client-retry-attempt-count=0, Prefer=return=minimal, x-ms-documentdb-partitionkey=[\"id28\"], x-ms-remaining-time-in-ms-on-client=
60000, Content-Type=application/json]","cosmosDiagnostics":{"userAgent":"azsdk-java-cosmos/4.12.0 Windows10/10.0 JRE/11.0.7","requestLatencyInMs":5377,"requestStartTimeUTC":"2021-03-16T17:17:52.294220100Z","requestEndTimeUTC":"2021-03-16T17:17:57.672023700Z","connectionMode":"DIRECT","responseStatis
ticsList":[{"storeResult":{"storePhysicalAddress":"rntbd://10.10.10.16:10253/apps/DocDbApp/services/DocDbServer21/partitions/a4cb4961-38c8-11e6-8106-8cdcd42c33be/replicas/1p/","lsn":0,"globalCommittedLsn":-1,"partitionKeyRangeId":null,"isValid":true,"statusCode":410,"subStatusCode":0,"isGone":true,"
isNotFound":false,"isInvalidPartition":false,"requestCharge":0.0,"itemLSN":-1,"sessionToken":null,"exception":"AsyncRntbdRequestRecord({\"args\":{\"transportRequestId\":5,\"activityId\":\"8a37e431-867b-11eb-92fa-8b46417fc6ba\",\"origin\":\"rntbd://10.10.10.16:10253\",\"replicaPath\":\"/apps/DocDbApp
/services/DocDbServer21/partitions/a4cb4961-38c8-11e6-8106-8cdcd42c33be/replicas/1p\",\"timeCreated\":\"2021-03-16T17:17:52.531586400Z\",\"lifetime\":\"PT5.1371635S\"},\"requestLength\":333714,\"responseLength\":-1,\"status\":{\"done\":false,\"cancelled\":false,\"completedExceptionally\":false},\"ti
meline\":[{\"eventName\":\"created\",\"startTimeUTC\":\"2021-03-16T17:17:52.531586400Z\",\"durationInMicroSec\":0},{\"eventName\":\"queued\",\"startTimeUTC\":\"2021-03-16T17:17:52.531586400Z\",\"durationInMicroSec\":0},{\"eventName\":\"channelAcquisitionStarted\",\"startTimeUTC\":\"2021-03-16T17:17:
52.531586400Z\",\"durationInMicroSec\":68816},{\"eventName\":\"pipelined\",\"startTimeUTC\":\"2021-03-16T17:17:52.600402800Z\",\"durationInMicroSec\":565488},{\"eventName\":\"transitTime\",\"startTimeUTC\":\"2021-03-16T17:17:53.165891600Z\",\"durationInMicroSec\":4503111},{\"eventName\":\"received\"
,\"startTimeUTC\":null,\"durationInMicroSec\":0},{\"eventName\":\"completed\",\"startTimeUTC\":null,\"durationInMicroSec\":0}]})","transportRequestTimeline":[{"eventName":"created","startTimeUTC":"2021-03-16T17:17:52.531586400Z","durationInMicroSec":0},{"eventName":"queued","startTimeUTC":"2021-03-1
6T17:17:52.531586400Z","durationInMicroSec":0},{"eventName":"channelAcquisitionStarted","startTimeUTC":"2021-03-16T17:17:52.531586400Z","durationInMicroSec":68816},{"eventName":"pipelined","startTimeUTC":"2021-03-16T17:17:52.600402800Z","durationInMicroSec":565488},{"eventName":"transitTime","startT
imeUTC":"2021-03-16T17:17:53.165891600Z","durationInMicroSec":4506132},{"eventName":"received","startTimeUTC":null,"durationInMicroSec":0},{"eventName":"completed","startTimeUTC":"2021-03-16T17:17:57.672023700Z","durationInMicroSec":0}],"rntbdRequestLengthInBytes":333714,"rntbdResponseLengthInBytes"
:-1,"requestPayloadLengthInBytes":333359,"responsePayloadLengthInBytes":null,"channelTaskQueueSize":1,"pendingRequestsCount":1,"serviceEndpointStatistics":{"availableChannels":0,"acquiredChannels":0,"executorTaskQueueSize":0,"inflightRequests":3,"lastSuccessfulRequestTime":"2021-03-16T17:17:52.530Z"
,"lastRequestTime":"2021-03-16T17:17:52.531Z","createdTime":"2021-03-16T17:17:52.529591500Z","isClosed":false}},"requestResponseTimeUTC":"2021-03-16T17:17:57.672023700Z","requestResourceType":"Document","requestOperationType":"Create"}],"supplementalResponseStatisticsList":[],"addressResolutionStati
stics":{"8d6d1e86-867b-11eb-92fa-8b46417fc6ba":{"startTimeUTC":"2021-03-16T17:17:57.677979600Z","endTimeUTC":null,"targetEndpoint":"https://10.10.10.16:8081/addresses/?$resolveFor=dbs%2FKbF7AA%3D%3D%2Fcolls%2FKbF7AOzF%2Bog%3D%2Fdocs&$filter=protocol%20eq%20rntbd&$partitionKeyRangeIds=0","errorMessag
e":null,"inflightRequest":true}},"regionsContacted":["https://10.10.10.16:8081/"],"retryContext":{"retryCount":0,"statusAndSubStatusCodes":null,"retryLatency":0},"metadataDiagnosticsContext":{"metadataDiagnosticList":[{"metaDataName":"PARTITION_KEY_RANGE_LOOK_UP","startTimeUTC":"2021-03-16T17:17:52.
340149600Z","endTimeUTC":"2021-03-16T17:17:52.434844Z","durationinMS":94}]},"serializationDiagnosticsContext":{"serializationDiagnosticsList":[{"serializationType":"ITEM_SERIALIZATION","startTimeUTC":"2021-03-16T17:17:52.294220100Z","endTimeUTC":"2021-03-16T17:17:52.295218300Z","durationInMicroSec":
998},{"serializationType":"PARTITION_KEY_FETCH_SERIALIZATION","startTimeUTC":"2021-03-16T17:17:52.321148300Z","endTimeUTC":"2021-03-16T17:17:52.324140200Z","durationInMicroSec":2991}]},"gatewayStatistics":null,"systemInformation":{"usedMemory":"1050664 KB","availableMemory":"1021912 KB","systemCpuLo
ad":"(2021-03-16T17:17:54.872327600Z 54,3%)"},"clientCfgs":{"id":0,"numberOfClients":1,"connCfg":{"rntbd":"(cto:PT5S, rto:PT5S, icto:PT0S, ieto:PT1H, mcpe:130, mrpc:30, cer:false)","gw":"(cps:1000, rto:PT5S, icto:null, p:false)","other":"(ed: true, cs: false)"},"consistencyCfg":"(consistency: null, 
mm: true, prgns: [])"}}}
	at com.azure.cosmos.implementation.directconnectivity.rntbd.RntbdRequestRecord.expire(RntbdRequestRecord.java:229)
	at io.netty.util.concurrent.DefaultEventExecutor.run(DefaultEventExecutor.java:66)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.base/java.lang.Thread.run(Thread.java:834)
	Suppressed: java.lang.Exception: #block terminated with an error
		at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:99)
		at reactor.core.publisher.Mono.block(Mono.java:1685)
		at TestCosmosBug.main(TestCosmosBug.java:45)
```

**Comment by xinlian12** [comment]
Please upgrade to the latest SDK, the issue should have been fixed >= 4.14.0  https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/cosmos/azure-cosmos/CHANGELOG.md#bugs-fixed-29
